smolyay/basis.py

```diff
@@ -403,29 +476,315 @@ def _derivative(self, x):
+                for b in basis_functions[1:]
+            ):
+                raise TypeError(
+                    "Basis functions must be be BasisFunction objects with the same domain and typing."
+                )
+        self._basis_functions = list(basis_functions)
 
     @property
     def basis_functions(self):
         """list: Basis functions."""
         return self._basis_functions
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain of the `basis_functions`"""
+        if len(self._basis_functions) == 0:
+            raise AttributeError("No basis functions to derive a domain.")
+        return self._basis_functions[0].domain
+
+    def __len__(self):
+        return len(self._basis_functions)
+
+    def __getitem__(self, key):
+        return self.basis_functions[key]
+
+    def scale_to_domain(self, points, old_domain):
+        if len(self._basis_functions) == 0:
+            raise AttributeError("No basis functions to derive a domain.")
+        return self.basis_functions[0].scale_to_domain(points, old_domain)
+
+    def __call__(self, X, domain=None):
+        """Evaluate all the basis functions in the set
+
+        Calls all the basis function(s) at index and evaluates at X.
+
+        Parameters
+        ----------
+        X : array-like
+            the points to evaluate.
+
+        domain : numpy array of shape (2,)
+            the lower and upper bounds of X.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+        if domain is not None:
+            new_X = self.scale_to_domain(X, domain)
+        else:
+            new_X = numpy.asarray(X)
+        y = numpy.zeros(
+            [len(self)] + list(new_X.shape),
+            dtype=complex if any(bf._is_complex for bf in self) else float,
+        )
+        for i, bf in enumerate(self._basis_functions):
+            y[i] = bf(new_X)
+        return y
+
+    def derivative(self, X, domain=None, n=1):
+        """Evaluate all the derivative of basis functions in the set
+
+        Calls the derivative for all the basis function(s) and
+        evaluates at X.
+
+        Parameters
+        ----------
+        X : array-like
+            the points to evaluate.
+
+        n : int, optional
+            order of derivative. Default is 1.
+
+        domain : numpy array of shape (2,)
+            the lower and upper bounds of X.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+        new_X = self.scale_to_domain(X, domain)
+        y = numpy.zeros(
+            [len(self)] + list(new_X.shape),
+            dtype=complex if any(bf._is_complex for bf in self) else float,
+        )
+        for i in range(len(self)):
+            y[i, :] = self[i].derivative(new_X, n)
+        y *= ((self.domain[1] - self.domain[0]) / (domain[1] - domain[0])) ** n
```

Comment:
This should only be done when domain is specified, right? Otherwise it will raise an exception that domain can't be indexed.
```suggestion
        if domain is not None:
            y *= ((self.domain[1] - self.domain[0]) / (domain[1] - domain[0])) ** n
```